modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.C

```diff
@@ -0,0 +1,610 @@
+    // std::cout << "  and the updated initial lattice friction value is " << _initial_lattice_friction(i) << "\n";
+  }
+}
+
+void
+CrystalPlasticityHCPDislocationSlipBeyerleinUpdate::setInitialConstitutiveVariableValues()
+{
+  _slip_resistance[_qp] = _slip_resistance_old[_qp];
+  _previous_substep_slip_resistance = _slip_resistance_old[_qp];
+
+  _forest_dislocation_density[_qp] = _forest_dislocation_density_old[_qp];
+  _previous_substep_forest_dislocations = _forest_dislocation_density_old[_qp];
+
+  _total_substructure_density[_qp] = _total_substructure_density_old[_qp];
+  _previous_substep_total_substructure_density = _total_substructure_density_old[_qp];
+}
+
+void
+CrystalPlasticityHCPDislocationSlipBeyerleinUpdate::setSubstepConstitutiveVariableValues()
+{
+  _slip_resistance[_qp] = _previous_substep_slip_resistance;
+  _forest_dislocation_density[_qp] = _previous_substep_forest_dislocations;
+  _total_substructure_density[_qp] = _previous_substep_total_substructure_density;
+}
+
+bool
+CrystalPlasticityHCPDislocationSlipBeyerleinUpdate::calculateSlipRate()
+{
+  // std::cout << "At the beginning of the calculateSlipRate method: \n";
+  for (unsigned int i = 0; i < _number_slip_systems; ++i)
+  {
+    // std::cout << "On slip system number " << i << ":\n";
+    // std::cout << "   the applied resolved stress is " << _tau[_qp][i] << "\n";
+    // std::cout << "   the slip resistance is " << _slip_resistance[_qp][i] << "\n";
+    // std::cout << "   and the reference strain rate value is " << _reference_strain_rate << "\n";
+    Real driving_force = std::abs(_tau[_qp][i] / _slip_resistance[_qp][i]);
+    _slip_increment[_qp][i] =
+        _reference_strain_rate * std::pow(driving_force, (1.0 / _rate_sensitivity_exponent));
+    if (_tau[_qp][i] < 0.0)
+      _slip_increment[_qp][i] *= -1.0;
+
+    // std::cout << "  then the calculated slip increment value is: " << _slip_increment[_qp][i] << "\n\n";
+
+    if (std::abs(_slip_increment[_qp][i]) * _substep_dt > _slip_incr_tol)
+    {
+#ifdef DEBUG
+      mooseWarning("Maximum allowable slip increment exceeded ",
+                   std::abs(_slip_increment[_qp][i]) * _substep_dt);
+#endif
+      return false;
+    }
+  }
+  return true;
+}
+
+void
+CrystalPlasticityHCPDislocationSlipBeyerleinUpdate::calculateConstitutiveSlipDerivative(
+    std::vector<Real> & dslip_dtau)
+{
+  calculateSlipDerivative(dslip_dtau);
+}
+
+void
+CrystalPlasticityHCPDislocationSlipBeyerleinUpdate::calculateSlipDerivative(
+    std::vector<Real> & ddislocaitonslip_dtau)
+{
+  for (unsigned int i = 0; i < _number_slip_systems; ++i)
+  {
+    if (MooseUtils::absoluteFuzzyEqual(_tau[_qp][i], 0.0))
+      ddislocaitonslip_dtau[i] = 0.0;
+    else
+      ddislocaitonslip_dtau[i] = _slip_increment[_qp][i] /
+                                 (_rate_sensitivity_exponent * std::abs(_tau[_qp][i])) *
+                                 _substep_dt;
+  }
+}
```

Comment:
Unify these two?

Reply:
I was trying to be a bit "forward-thinking" here: when I add twinning as a separate inheriting class, I was hoping to overwrite the `calculateConstitutiveSlipDerivative` method and then still call the `calculateSlipDerivative` method. Is there a better way to have this flexibility for the child class in a unified method?